### ISSUE TITLE
A bunch of native dependencies are blowing us up

### ISSUE THREAD

**Issue by yaronyg** [thaliproject/Thali_CordovaPlugin#926]
See https://gist.github.com/yaronyg/83873df17c44238373f9cb8c56e65aec for the end of the terminal output when I try to run 'jx npm run setupDesktop' in Thali_CordovaPlugin/thali/install on vNext_yarong_417_1.

Note that I deleted all my node_modules, my .jx directory and also my .node-gyp directory before doing the build.

I then tried to build vNext and it built just fine. I noticed in the error log that we are trying to build leveldown@1.4.6 in vNext_yarong_417_1 which we are not trying to build in vNext.

So I removed the use of pouchdb-node which now seems to mandate leveldown and replaced it with pouchdb where leveldown is optional (but the size of the module is larger).

However next I now have blow ups in TestServer, we have dependencies on bufferutil and utf-8-validate.

So these are optional dependencies that are brought in by socket.io. So for now I need to update our install script to use --no-optional. But I also need to file a bug that JXcore npm install should gracefully fail out if it can't build a native optional component. And no, we aren't ever going to implement that. But I'll file it anyway.

**Comment by yaronyg** [comment]
Meanwhile I now have a blow up because I'm getting an error about leveldown not being available! This is coming from testThaliReplicationPeerAction/Make sure docs replicate. I fixed this by modifying getPouchDBFActoryInRandomDirectory to point to leveldown-mobile as its default adapter.

But that just led me to a problem I swore we had fixed, which is that the tests hang at the end because something isn't exiting where it should on 'jx npm test' . And when I ran 'jx npm run test-coordinated' I got an ETIMEDOUT error. Both of these errors are exactly why I had to build our own custom releases of PouchDB and Express-PouchDB. Which I suppose makes sense as vNext_yarong_417_1 doesn't have the changes made to support that.

So now I need to take my changes back up to thaliproject/Thali_CordovaPlugin#893 and see if I can get that to behave itself and pass. Oh the joy.

**Comment by yaronyg** [comment]
I created a new branch vNext_yarong_417_812 based on the changes I made above to vNext_yarong_417_1 and I then merge vNext_yarong_812 into that branch (somehow merging in that direction just made more sense to me). So now I'm dealing with the problem that installCustomPouchDB wants to create pouchdb-node. But per #870 we were going to get rid of that work anyway so I basically just implemented #870 myself.

However now the ETIMEDOUT error is back! I caught it in 'Coordinated replication action test' inside of testThaliReplicationPeerActionCoordinated. So I'm trying to run the coordinated tests without it to see if it just that test or if it's something more systemic.So when I run without the test everything passes. When I run the test by itself everything passes. I did another run without that test and still got a ETIMEDOUT. It looks like our PouchDB bug is back. :(

I don't see much of a choice. I'm going to have to chase down this error. I'll try to use longjohn (and yes we have to use no-optional when installing it). And... um... just requiring longjohn causes my code to exit immediately. No error. Nothing. Just.... exit. So something odd happens. I'm making these changes to runCoordinatedTests and if I include longjohn and then require child_process that is fine but after I require minimist execution just stops.

I'm not sure what's going on here so I'm going to skip longjohn for the second and instead use bluebird. I can connect that right to pouchdb. Except once I put in Bluebird into PouchDB then I'm getting a new fun error:

` ` `
dyld: lazy symbol binding failed: Symbol not found: _node_module_register
  Referenced from: /Users/yarongoland/Desktop/depots/Thali_CordovaPlugin/test/www/jxcore/node_modules/leveldown/build/Release/leveldown.node
  Expected in: flat namespace
` ` `

This is going to be one of those days, isn't it?

Oh it gets better. This isn't just from Bluebird. It happens without it. I suspect I screwed up and at some point used npm instead of jx install and now I'm in a bad state. Time to nuke everything from orbit. :(

I'll give longjohn another try just incase my screw up with npm caused the issue.

O.k. you know I need to be more careful with my time. I think the right fix here is to cancel 870 and use our custom release of PouchDB. Except that just brings back a different bug that Andrew ran into in our custom release of PouchDB. So I suspect I really am stuck and I need to just find this ETIMEDOUT bug so we can use the latest release of PouchDB.

**Comment by yaronyg** [comment]
So now I'm trying 'jx npm install longjohn --no-optional'. And I'm going to add require('longjohn') to the top of runCoordinatedTests.js and I get the same behavior. It exits out without doing anything. An interesting point btw is that putting longjohn into runCoordinatedTests is pretty useless since all the interest stuff happens in the spawned processes but whatever. It seems like longjohn doesn't like Spidermonkey (is my guess).

So now I'll try 'jx npm install bluebird --no-optional' and now I have to go find the index.js for PouchDB and manually point its PouchDBPromise at bluebird and configure bluebird to issue long stacks. I installed bluebird but before connecting it I am running a coordinated test just to make sure that the dyld error doesn't show up again. Not only didn't the dyld error show up again but the coordinated tests passed. I'll try again just to make sure it wasn't luck. And now it just passed again. O.k. one more try and then I'm declaring it working. (Just typing that should be enough to cause a failure). Nope. It passed. O.k. I call it good!

Now I need to do a mobile build just to make sure I didn't break that. Unfortunately I did. But the error is really seriously weird:

Copying files from /Users/yarongoland/Desktop/depots/Thali_CordovaPlugin to /Users/yarongoland/Desktop/depots/ThaliTest/thaliDontCheckIn/Thali_CordovaPlugin-npmv2.1.0
Failed with - Error: copyDevelopmentThaliCordovaPluginToProject failed withError: ENOENT, lstat '/Users/yarongoland/Desktop/depots/Thali_CordovaPlugin/.git/logs/refs/remotes/origin/Ã

What the heck is that? I'm going to flail around and assume this is because of the fact that I'm using an old version of jxcore-cordova. Which also sounds sorry to me too but I'm trying things here! And updating the version didn't fix anything. :( Not that I really expected it to. So now I cleared out all my node_modules because sometimes that has caused issues in the past. O.k. it's just broke. I'll have to try and fix it tomorrow.
